src/NHibernate/Engine/JoinSequence.cs

```diff
@@ -175,7 +175,7 @@ public JoinFragment ToJoinFragment(IDictionary<string, IFilter> enabledFilters,
 			{
 				Join join = joins[i];
 				string on = join.AssociationType.GetOnCondition(join.Alias, factory, enabledFilters);
```

Comment:
I initially though the `GetOnCondition` implementation may be troublesome in some cases, because it does filter `enabledFilters` according to their `use-many-to-one` setting:
https://github.com/nhibernate/nhibernate-core/blob/e99cf4378d7588c77fd13e487e8fa825675d23dd/src/NHibernate/Type/EntityType.cs#L501-L512

And later here, we take the `on` clause if it is not empty, instead of computing it with all filters if `ForceFilter`. It appears this `on` clause will always be empty in our case, arbitrary joins, because `GetOnCondition` current implementation always yields that excepted for property-ref associations.
https://github.com/nhibernate/nhibernate-core/blob/e99cf4378d7588c77fd13e487e8fa825675d23dd/src/NHibernate/Type/EntityType.cs#L496-L499

So, the current change is enough to fix the trouble. I cannot think of a related test case demonstrating a need to change `GetOnCondition`. This implementation looks a bit brittle to me, but a patch change is not the place to get things better from that viewpoint, unless we can demonstrate a failing case.